### PR TITLE
OOP school library: unit tests

### PR DESCRIPTION
**Pull request by juaryimami**
### In this project:-
---
 - We add unit tests to our code base, specifically to our classes.
#### Add unit tests for all the methods in all the classes:
- Book
- Person
- Student
- Teacher
- Classroom
- Rental
- all classes that you created while implementing the Decorator design pattern
---
#### Thank you for Reviewing my project.

### CODE REVIEW

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+require_relative '../classes/person'
+
+describe Person do
+  age = 22
+  name = 'Akuma'
+  let(:person) { Person.new age, name }
+
```
Comment (KDlamini) [inline spec/person_spec.rb:7]: This milestone requires that there are unit tests for all the public methods in all the classes.

- There are three methods in the `Person` class that are not tested. Kindly make sure that the `can_use_services?`,  `validate_name`, and `add_rental` methods are tested for the `Person` class.

File: spec/book_spec.rb
Comment (KDlamini) [inline spec/book_spec.rb:1]: All tests pass ✔️ well done!!! 🥳🎆 

image

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @juaryimami, @aubinleyoung 🙋‍♀️,  

Gret work so far! 💯 
You have done a great job writing the unit tests OOP school library. 🎊 
There are some issues that you still need to work on to improve your project according to the requirements. You are almost there! 💪🏾 

## Highlights 🌞

✔️  Good PR title and description
✔️  All test files are placed in a different directory (e.g `/spec`)
✔️  There is one file per class
✔️  there is an 'it' block for each class method

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @juaryimami, @aubinleyoung 🙋‍♀️,  

Gret work so far! 💯 
You have done a great job writing the unit tests OOP school library. 🎊 
There are some issues that you still need to work on to improve your project according to the requirements. You are almost there! 💪🏾 

## Highlights 🌞

✔️  Good PR title and description
✔️  All test files are placed in a different directory (e.g `/spec`)
✔️  There is one file per class
✔️  there is an 'it' block for each class method

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/person_spec.rb (hunk 1 above)
Comment (juaryimami) [reply spec/person_spec.rb:7]: ok thank you I will add those soon

File: spec/person_spec.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  it 'returns the correct name' do
+    expect(person.name).to eq(name)
+  end
+  it 'creates a new Person object with default values' do
+    expect(person.parent_permission).to eq(true)
+    expect(person.can_use_services?).to eq(true)
+    expect(person.rentals).to eq([])
+  end
```
Comment (Shedrack-Sunday) [inline spec/person_spec.rb:23]: - [  ] Good job taking heed from the last reviewer 👏 . However, can we specifically test with the functions as specified in the requirements and as pointed out by the last reviewer?
- the functions `validate_name` and `add_rental` should be implemnted 

## Example:
```
    it "The validate_name method corrects the name" do
      person = Person.new(20, "nameTest")
      person.validate_name()
      expect(person.name).to eql "nameTest"
    end
    
    
it "The add_rental method returns a rental and adds its to the book" do
      book = Book.new("Title", "Author")
      person = Person.new(24, "Bob")

      rental = person.add_rental(book, "2021-05-04")

      expect(person.rentals).to eql [rental]
    end

```
## Requirements:
image

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @juaryimami    👋 

# CHANGES REQUIRED ♻️

Great work on the project 👏
You've done well implementing the requirements,  However, there are a few you need to work on, to move to the next project. 💯

 


## Highlights 
- Nice code structure✔️ 
- Nice use duck method 👏 
- clean implementiion f `Rspec` test✔️ 
  
# Required Changes ♻️

- [ ] You did great on the project, however, there are a few changes needed in testing the functions of the `Person` class

_check the comment section for a detailed review_ 

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/person_spec.rb (hunk 2 above)
Comment (juaryimami) [reply spec/person_spec.rb:23]: Thank you for your supportive feedback Sir.

**Review by pinkmoon25 — APPROVED** [review]
Hi @juaryimami @aubinleyoung , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
Rubocop checks :heavy_check_mark: 
gitflow :heavy_check_mark: 
changes implemented :heavy_check_mark: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**